File: src/catkin_ws/src/spot_keyword_spotting/scripts/i2s_window_node.py
#!/usr/bin/env python3
# spot_keyword_spotting/scripts/i2s_window_node.py

from __future__ import annotations

import errno
import time

import alsaaudio
import numpy as np
import rospy

from std_msgs.msg import Float32MultiArray, MultiArrayDimension


SAMPLE_FORMATS = {
    "S16_LE": (alsaaudio.PCM_FORMAT_S16_LE, 2, np.dtype("<i2"), 32768.0),
    "S32_LE": (alsaaudio.PCM_FORMAT_S32_LE, 4, np.dtype("<i4"), 2147483648.0),
}


class AveragingResampler:
    def __init__(self, input_sample_rate: int, output_sample_rate: int) -> None:
        self.input_sample_rate = int(input_sample_rate)
        self.output_sample_rate = int(output_sample_rate)
        if self.input_sample_rate <= 0 or self.output_sample_rate <= 0:
            raise ValueError("sample rates must be positive")
        if self.input_sample_rate % self.output_sample_rate != 0:
            raise ValueError("input sample rate must be an integer multiple of output sample rate")
        self.factor = self.input_sample_rate // self.output_sample_rate
        self.pending = np.empty(0, dtype=np.float32)
        self.input_count = 0
        self.output_count = 0

    def process(self, audio: np.ndarray) -> np.ndarray:
        audio = np.asarray(audio, dtype=np.float32).reshape(-1)
        if audio.size == 0:
            return np.empty(0, dtype=np.float32)

        self.input_count += int(audio.size)
        if self.factor == 1:
            self.output_count += int(audio.size)
            return audio.astype(np.float32, copy=False)

        if self.pending.size:
            audio = np.concatenate((self.pending, audio))

        usable = audio.size - (audio.size % self.factor)
        self.pending = audio[usable:].copy()
        if usable <= 0:
            return np.empty(0, dtype=np.float32)

        out = audio[:usable].reshape(-1, self.factor).mean(axis=1).astype(np.float32)
        self.output_count += int(out.size)
        return out


class RingWindowPublisher:
    def __init__(self, window_samples: int, hop_samples: int, sample_rate: int, publisher: rospy.Publisher) -> None:
        self.window_samples = int(window_samples)
        self.hop_samples = int(hop_samples)
        self.sample_rate = int(sample_rate)
        self.publisher = publisher
        self.ring = np.zeros(self.window_samples, dtype=np.float32)
        self.write_index = 0
        self.samples_seen = 0
        self.samples_since_publish = 0

    @property
    def ready(self) -> bool:
        return self.samples_seen >= self.window_samples

    def push(self, audio: np.ndarray) -> None:
        audio = np.asarray(audio, dtype=np.float32).reshape(-1)
        for sample in audio:
            self.ring[self.write_index] = sample
            self.write_index = (self.write_index + 1) % self.window_samples
            self.samples_seen += 1
            self.samples_since_publish += 1
            if self.ready and self.samples_since_publish >= self.hop_samples:
                self.publish()
                self.samples_since_publish = 0

    def current_window(self) -> np.ndarray:
        return np.concatenate((self.ring[self.write_index:], self.ring[:self.write_index])).astype(np.float32)

    def publish(self) -> None:
        window = self.current_window()
        msg = Float32MultiArray()
        msg.layout.dim = [
            MultiArrayDimension("sample_rate", self.sample_rate, self.sample_rate),
            MultiArrayDimension("window_samples", self.window_samples, self.window_samples),
            MultiArrayDimension("samples_seen", min(int(self.samples_seen), 2**32 - 1), 1),
        ]
        msg.data = window.tolist()
        self.publisher.publish(msg)
        rospy.loginfo_throttle(
            5.0,
            "Published Python KWS audio window: samples=%d, hop=%d, samples_seen=%d",
            window.size,
            self.hop_samples,
            self.samples_seen,
        )


class I2SWindowNode:
    def __init__(self) -> None:
        self.device = rospy.get_param("~device", "hw:1,0")
        self.input_sample_rate = int(rospy.get_param("~input_sample_rate", 48000))
        self.output_sample_rate = int(rospy.get_param("~output_sample_rate", 16000))
        self.channels = int(rospy.get_param("~channels", 2))
        self.period_size = int(rospy.get_param("~period_size", 512))
        self.sample_format = str(rospy.get_param("~sample_format", "S32_LE")).upper()
        self.channel_index = int(rospy.get_param("~channel_index", 0))
        self.audio_gain = float(rospy.get_param("~audio_gain", 1.0))
        self.window_samples = int(rospy.get_param("~window_samples", 16000))
        self.hop_samples = int(rospy.get_param("~hop_samples", 3200))
        self.window_topic = rospy.get_param("~window_topic", "/voice/audio_window")

        if self.sample_format not in SAMPLE_FORMATS:
            raise ValueError("unsupported sample_format {}; supported: {}".format(self.sample_format, sorted(SAMPLE_FORMATS)))
        self.alsa_format, self.bytes_per_sample, self.dtype, self.scale = SAMPLE_FORMATS[self.sample_format]

        self.publisher = rospy.Publisher(self.window_topic, Float32MultiArray, queue_size=1)
        self.resampler = AveragingResampler(self.input_sample_rate, self.output_sample_rate)
        self.window_publisher = RingWindowPublisher(
            self.window_samples,
            self.hop_samples,
            self.output_sample_rate,
            self.publisher,
        )
        self.audio_messages = 0
        self.audio_bytes = 0
        self.alsa_overruns = 0
        self.alsa_errors = 0
        self.last_recover_time = 0.0
        self.pcm = None
        self._open_pcm()

        rospy.loginfo(
            "Python I2S window node started: device=%s, in=%d Hz, out=%d Hz, channels=%d, format=%s, period=%d, channel=%d, window=%d, hop=%d, topic=%s",
            self.device,
            self.input_sample_rate,
            self.output_sample_rate,
            self.channels,
            self.sample_format,
            self.period_size,
            self.channel_index,
            self.window_samples,
            self.hop_samples,
            self.window_topic,
        )

    def _open_pcm(self) -> None:
        self.pcm = alsaaudio.PCM(alsaaudio.PCM_CAPTURE, alsaaudio.PCM_NORMAL, device=self.device)
        self.pcm.setchannels(self.channels)
        self.pcm.setrate(self.input_sample_rate)
        self.pcm.setformat(self.alsa_format)
        self.pcm.setperiodsize(self.period_size)

    def _recover_pcm(self) -> None:
        now = time.monotonic()
        if now - self.last_recover_time < 0.05:
            return
        self.last_recover_time = now
        try:
            if self.pcm is not None:
                self.pcm.close()
        except (AttributeError, alsaaudio.ALSAAudioError):
            pass
        self._open_pcm()

    def _read_mono(self) -> np.ndarray | None:
        try:
            length, data = self.pcm.read()
        except alsaaudio.ALSAAudioError as exc:
            self.alsa_errors += 1
            rospy.logwarn_throttle(2.0, "ALSA read failed in Python window node: %s (count=%d)", exc, self.alsa_errors)
            try:
                self._recover_pcm()
            except alsaaudio.ALSAAudioError as recover_exc:
                rospy.logwarn_throttle(2.0, "Failed to recover ALSA stream: %s", recover_exc)
            return None

        if length <= 0:
            if length == -errno.EPIPE:
                self.alsa_overruns += 1
                rospy.logwarn_throttle(2.0, "ALSA overrun in Python window node (count=%d); recovering", self.alsa_overruns)
                try:
                    self._recover_pcm()
                except alsaaudio.ALSAAudioError as recover_exc:
                    rospy.logwarn_throttle(2.0, "Failed to recover ALSA stream: %s", recover_exc)
                return None
            rospy.logwarn_throttle(2.0, "ALSA read returned length=%d", length)
            return None

        samples = np.frombuffer(data, dtype=self.dtype).astype(np.float32)
        expected_values = int(length) * self.channels
        if samples.size < expected_values:
            rospy.logwarn_throttle(2.0, "Short ALSA block: expected %d values, got %d", expected_values, samples.size)
            return None
        samples = samples[:expected_values].reshape(int(length), self.channels)

        if 0 <= self.channel_index < self.channels:
            selected_channel = self.channel_index
        else:
            rms_by_channel = np.sqrt(np.mean(samples * samples, axis=0))
            selected_channel = int(np.argmax(rms_by_channel))

        mono = samples[:, selected_channel] / self.scale
        mono = np.clip(mono * self.audio_gain, -1.0, 1.0).astype(np.float32)
        self.audio_messages += 1
        self.audio_bytes += len(data)
        rospy.loginfo_throttle(
            5.0,
            "Python I2S alive: messages=%d bytes=%d frames=%d selected_channel=%d peak=%.6f rms=%.6f",
            self.audio_messages,
            self.audio_bytes,
            length,
            selected_channel,
            float(np.max(np.abs(mono))) if mono.size else 0.0,
            float(np.sqrt(np.mean(mono * mono))) if mono.size else 0.0,
        )
        return mono

    def run(self) -> None:
        while not rospy.is_shutdown():
            audio_48k = self._read_mono()
            if audio_48k is None:
                continue
            audio_16k = self.resampler.process(audio_48k)
            self.window_publisher.push(audio_16k)
            rospy.loginfo_throttle(
                5.0,
                "Python window resampler: in_total=%d out_total=%d latest_in=%d latest_out=%d pending=%d",
                self.resampler.input_count,
                self.resampler.output_count,
                audio_48k.size,
                audio_16k.size,
                self.resampler.pending.size,
            )


def main() -> None:
    rospy.init_node("i2s_window_node")
    I2SWindowNode().run()


if __name__ == "__main__":
    main()

File: src/catkin_ws/src/spot_keyword_spotting/src/i2s_window_node.cpp
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <numeric>
#include <stdexcept>
#include <string>
#include <vector>

#include <alsa/asoundlib.h>
#include <ros/ros.h>
#include <std_msgs/Float32MultiArray.h>
#include <std_msgs/MultiArrayDimension.h>

class I2SWindowNode {
public:
  I2SWindowNode()
      : private_nh_("~") {
    private_nh_.param<std::string>("device", device_, "hw:1,0");
    private_nh_.param("input_sample_rate", input_sample_rate_, 48000);
    private_nh_.param("output_sample_rate", output_sample_rate_, 16000);
    private_nh_.param("channels", channels_, 2);
    private_nh_.param("period_size", period_size_, 512);
    private_nh_.param("channel_index", channel_index_, 0);
    private_nh_.param("audio_gain", audio_gain_, 1.0);
    private_nh_.param<std::string>("sample_format", sample_format_, "S32_LE");
    private_nh_.param<std::string>("window_topic", window_topic_, "/voice/audio_window");
    private_nh_.param("window_samples", window_samples_, 16000);
    private_nh_.param("hop_samples", hop_samples_, 3200);

    if (input_sample_rate_ <= 0 || output_sample_rate_ <= 0 || channels_ <= 0) {
      throw std::runtime_error("sample rates and channels must be positive");
    }
    if (input_sample_rate_ % output_sample_rate_ != 0) {
      throw std::runtime_error("input_sample_rate must be an integer multiple of output_sample_rate");
    }

    decimation_factor_ = input_sample_rate_ / output_sample_rate_;
    ring_.assign(static_cast<size_t>(window_samples_), 0.0f);

    window_pub_ = nh_.advertise<std_msgs::Float32MultiArray>(window_topic_, 2);
    openPcm();

    ROS_INFO("I2S window node started: device=%s, in=%d Hz, out=%d Hz, channels=%d, format=%s, period=%d, channel=%d, window=%d, hop=%d, topic=%s",
             device_.c_str(), input_sample_rate_, output_sample_rate_, channels_, sample_format_.c_str(), period_size_,
             channel_index_, window_samples_, hop_samples_, window_topic_.c_str());
  }

  ~I2SWindowNode() {
    if (pcm_ != nullptr) {
      snd_pcm_close(pcm_);
      pcm_ = nullptr;
    }
  }

  void run() {
    std::vector<uint8_t> bytes(static_cast<size_t>(period_size_ * channels_ * bytes_per_sample_));

    while (ros::ok()) {
      snd_pcm_sframes_t frames = snd_pcm_readi(pcm_, bytes.data(), period_size_);
      if (frames < 0) {
        handleReadError(frames);
        ros::spinOnce();
        continue;
      }
      if (frames == 0) {
        ros::spinOnce();
        continue;
      }

      processBlock(bytes.data(), static_cast<size_t>(frames));
      ros::spinOnce();
    }
  }

private:
  void openPcm() {
    if (pcm_ != nullptr) {
      snd_pcm_close(pcm_);
      pcm_ = nullptr;
    }

    int err = snd_pcm_open(&pcm_, device_.c_str(), SND_PCM_STREAM_CAPTURE, 0);
    if (err < 0) {
      throw std::runtime_error("snd_pcm_open failed: " + std::string(snd_strerror(err)));
    }

    snd_pcm_format_t format = SND_PCM_FORMAT_S32_LE;
    bytes_per_sample_ = 4;
    scale_ = 2147483648.0f;

    if (sample_format_ == "S16_LE") {
      format = SND_PCM_FORMAT_S16_LE;
      bytes_per_sample_ = 2;
      scale_ = 32768.0f;
    } else if (sample_format_ != "S32_LE") {
      throw std::runtime_error("unsupported sample_format: " + sample_format_);
    }

    snd_pcm_hw_params_t *params = nullptr;
    snd_pcm_hw_params_alloca(&params);
    snd_pcm_hw_params_any(pcm_, params);
    snd_pcm_hw_params_set_access(pcm_, params, SND_PCM_ACCESS_RW_INTERLEAVED);
    snd_pcm_hw_params_set_format(pcm_, params, format);
    snd_pcm_hw_params_set_channels(pcm_, params, static_cast<unsigned int>(channels_));

    unsigned int rate = static_cast<unsigned int>(input_sample_rate_);
    snd_pcm_hw_params_set_rate_near(pcm_, params, &rate, nullptr);
    snd_pcm_uframes_t period = static_cast<snd_pcm_uframes_t>(period_size_);
    snd_pcm_hw_params_set_period_size_near(pcm_, params, &period, nullptr);

    int err2 = snd_pcm_hw_params(pcm_, params);
    if (err2 < 0) {
      throw std::runtime_error("snd_pcm_hw_params failed: " + std::string(snd_strerror(err2)));
    }

    snd_pcm_prepare(pcm_);
  }

  void handleReadError(snd_pcm_sframes_t frames) {
    if (frames == -EPIPE) {
      ++overruns_;
      ROS_WARN_THROTTLE(2.0, "ALSA overrun in C++ capture (count=%d); preparing stream", overruns_);
      snd_pcm_prepare(pcm_);
      return;
    }

    ROS_WARN_THROTTLE(2.0, "ALSA read failed: %s", snd_strerror(static_cast<int>(frames)));
    snd_pcm_prepare(pcm_);
  }

  void processBlock(const uint8_t *data, size_t frames) {
    latest_peak_ = 0.0f;
    latest_rms_accum_ = 0.0;

    for (size_t frame = 0; frame < frames; ++frame) {
      float sample = readSelectedSample(data, frame);
      sample = std::max(-1.0f, std::min(1.0f, sample * static_cast<float>(audio_gain_)));
      latest_peak_ = std::max(latest_peak_, std::abs(sample));
      latest_rms_accum_ += static_cast<double>(sample) * sample;
      pushInputSample(sample);
    }

    total_input_frames_ += frames;
    ROS_INFO_THROTTLE(5.0, "C++ I2S alive: in_frames=%llu out_samples=%llu latest_frames=%zu peak=%.6f rms=%.6f",
                      static_cast<unsigned long long>(total_input_frames_),
                      static_cast<unsigned long long>(total_output_samples_),
                      frames,
                      latest_peak_,
                      std::sqrt(latest_rms_accum_ / std::max<size_t>(1, frames)));
  }

  float readSelectedSample(const uint8_t *data, size_t frame) const {
    int selected = channel_index_;
    if (selected < 0 || selected >= channels_) {
      selected = 0;
    }

    const size_t offset = (frame * static_cast<size_t>(channels_) + static_cast<size_t>(selected)) * bytes_per_sample_;
    if (bytes_per_sample_ == 2) {
      int16_t raw = 0;
      std::memcpy(&raw, data + offset, sizeof(raw));
      return static_cast<float>(raw) / scale_;
    }

    int32_t raw = 0;
    std::memcpy(&raw, data + offset, sizeof(raw));
    return static_cast<float>(raw) / scale_;
  }

  void pushInputSample(float sample) {
    decimation_accum_ += sample;
    ++decimation_count_;

    if (decimation_count_ < decimation_factor_) {
      return;
    }

    const float out = decimation_accum_ / static_cast<float>(decimation_factor_);
    decimation_accum_ = 0.0f;
    decimation_count_ = 0;
    pushOutputSample(out);
  }

  void pushOutputSample(float sample) {
    ring_[ring_write_index_] = sample;
    ring_write_index_ = (ring_write_index_ + 1) % ring_.size();
    ++total_output_samples_;
    ++samples_since_publish_;
    if (!ring_full_ && total_output_samples_ >= static_cast<uint64_t>(ring_.size())) {
      ring_full_ = true;
    }

    if (ring_full_ && samples_since_publish_ >= static_cast<uint64_t>(hop_samples_)) {
      publishWindow();
      samples_since_publish_ = 0;
    }
  }

  void publishWindow() {
    std_msgs::Float32MultiArray msg;
    msg.layout.dim.resize(3);
    msg.layout.dim[0].label = "sample_rate";
    msg.layout.dim[0].size = static_cast<uint32_t>(output_sample_rate_);
    msg.layout.dim[0].stride = static_cast<uint32_t>(output_sample_rate_);
    msg.layout.dim[1].label = "window_samples";
    msg.layout.dim[1].size = static_cast<uint32_t>(window_samples_);
    msg.layout.dim[1].stride = static_cast<uint32_t>(window_samples_);
    msg.layout.dim[2].label = "samples_seen";
    msg.layout.dim[2].size = static_cast<uint32_t>(std::min<uint64_t>(total_output_samples_, UINT32_MAX));
    msg.layout.dim[2].stride = 1;
    msg.data.resize(ring_.size());

    const size_t start = ring_write_index_;
    for (size_t i = 0; i < ring_.size(); ++i) {
      msg.data[i] = ring_[(start + i) % ring_.size()];
    }

    window_pub_.publish(msg);
    ROS_INFO_THROTTLE(5.0, "Published KWS audio window: samples=%zu, hop=%d", msg.data.size(), hop_samples_);
  }

  ros::NodeHandle nh_;
  ros::NodeHandle private_nh_;
  ros::Publisher window_pub_;
  snd_pcm_t *pcm_ = nullptr;

  std::string device_;
  std::string sample_format_;
  std::string window_topic_;
  int input_sample_rate_ = 48000;
  int output_sample_rate_ = 16000;
  int channels_ = 2;
  int period_size_ = 512;
  int channel_index_ = 0;
  int window_samples_ = 16000;
  int hop_samples_ = 3200;
  int decimation_factor_ = 3;
  int decimation_count_ = 0;
  int bytes_per_sample_ = 4;
  int overruns_ = 0;
  double audio_gain_ = 1.0;
  float scale_ = 2147483648.0f;
  float decimation_accum_ = 0.0f;
  float latest_peak_ = 0.0f;
  double latest_rms_accum_ = 0.0;

  std::vector<float> ring_;
  size_t ring_write_index_ = 0;
  bool ring_full_ = false;
  uint64_t total_input_frames_ = 0;
  uint64_t total_output_samples_ = 0;
  uint64_t samples_since_publish_ = 0;
};

int main(int argc, char **argv) {
  ros::init(argc, argv, "i2s_window_node");

  try {
    I2SWindowNode node;
    node.run();
  } catch (const std::exception &exc) {
    ROS_FATAL("i2s_window_node failed: %s", exc.what());
    return 1;
  }

  return 0;
}